build: Re-enable external signer on Windows

hebasto opened the pull request:
As https://github.com/boostorg/process/issues/207 has been resolved, it is possible now to re-enable external signer on Windows when cross-compiling.

Guix build hashes:
```
78f69ea7e0dbc8338981a92c0352220ccd7c2272d8cbff6a3b082a1412a935c5  guix-build-1a0d8e178c7b/output/aarch64-linux-gnu/SHA256SUMS.part
ee17456ec818ddf5a175182508966e622573ccb518807cca43a40fa1dceda092  guix-build-1a0d8e178c7b/output/aarch64-linux-gnu/bitcoin-1a0d8e178c7b-aarch64-linux-gnu-debug.tar.gz
5080551bde379c746cc67b10429aef33b9f9e49d2d4e21ee1c3bfd9c1c845d46  guix-build-1a0d8e178c7b/output/aarch64-linux-gnu/bitcoin-1a0d8e178c7b-aarch64-linux-gnu.tar.gz
dfab220ce76a40bf7dcf07aab352a616a91b516503639455fe7e1b137bad3e85  guix-build-1a0d8e178c7b/output/arm-linux-gnueabihf/SHA256SUMS.part
516ceb822571a8bd88fe107dca434ef596b1e4328ccbda1d51e1d482d3050396  guix-build-1a0d8e178c7b/output/arm-linux-gnueabihf/bitcoin-1a0d8e178c7b-arm-linux-gnueabihf-debug.tar.gz
21325380638f817107c203b9a1aedb808d1a4a2b4041493753ca4cbf19aa4f2c  guix-build-1a0d8e178c7b/output/arm-linux-gnueabihf/bitcoin-1a0d8e178c7b-arm-linux-gnueabihf.tar.gz
cf48ed78fcfceaeb3610ccf22326d735a129dcbf9d50b557b3de359169aefdfd  guix-build-1a0d8e178c7b/output/arm64-apple-darwin/SHA256SUMS.part
d4d51e136148bac6a20bb3adb402c499967647736acb420bfdeb71603aba57da  guix-build-1a0d8e178c7b/output/arm64-apple-darwin/bitcoin-1a0d8e178c7b-arm64-apple-darwin-unsigned.dmg
95bb62d24f860e08a392ddb74d5860ccf27e8baa183e6749af877d26a3bd6b0b  guix-build-1a0d8e178c7b/output/arm64-apple-darwin/bitcoin-1a0d8e178c7b-arm64-apple-darwin-unsigned.tar.gz
68da4c92f37bb802df37141af194f47c16da1d84f77a0fbb1016013ae0338502  guix-build-1a0d8e178c7b/output/arm64-apple-darwin/bitcoin-1a0d8e178c7b-arm64-apple-darwin.tar.gz
6704e38c2d3f11321403797598d05f062648fec6f2d76900ba250dab481e29da  guix-build-1a0d8e178c7b/output/dist-archive/bitcoin-1a0d8e178c7b.tar.gz
64b936bc90d1e01fe8f276511edc9bb945dcebe70332aa37d3a786348443b8e7  guix-build-1a0d8e178c7b/output/powerpc64-linux-gnu/SHA256SUMS.part
3d03532e54b6e42498ea240c86b8567e94fd462f56087b869c3d6f09e2dde878  guix-build-1a0d8e178c7b/output/powerpc64-linux-gnu/bitcoin-1a0d8e178c7b-powerpc64-linux-gnu-debug.tar.gz
c5843d79a58b0a864fe723458dab4eee54ad11f4b1f7960975b086eeedc0d541  guix-build-1a0d8e178c7b/output/powerpc64-linux-gnu/bitcoin-1a0d8e178c7b-powerpc64-linux-gnu.tar.gz
f861ff519bd5e3d6d5ce1646ee0a06bcef1288ddb804a4a600e4dbfe5d5be521  guix-build-1a0d8e178c7b/output/powerpc64le-linux-gnu/SHA256SUMS.part
5f477da21980dbcf9696081903dc1ba8a3f79ce3579641d208e69a6f598c8eb9  guix-build-1a0d8e178c7b/output/powerpc64le-linux-gnu/bitcoin-1a0d8e178c7b-powerpc64le-linux-gnu-debug.tar.gz
b3757b11c614136934158acea5139e8abd0c5c9cdfda72ae44db436f21716b33  guix-build-1a0d8e178c7b/output/powerpc64le-linux-gnu/bitcoin-1a0d8e178c7b-powerpc64le-linux-gnu.tar.gz
1c21bdb17fe3436e685e88c62423e630fe2b3c41dd00025a99fd80d97817ac2f  guix-build-1a0d8e178c7b/output/riscv64-linux-gnu/SHA256SUMS.part
f36ae98473f086ae8f0dc66223b5ec407d57dc4d8d45ae284401520ff5c0b273  guix-build-1a0d8e178c7b/output/riscv64-linux-gnu/bitcoin-1a0d8e178c7b-riscv64-linux-gnu-debug.tar.gz
1603e4d0e869eb47a1dc2d26b67772d0016d90f7ba5e50d2009365cc02cb8169  guix-build-1a0d8e178c7b/output/riscv64-linux-gnu/bitcoin-1a0d8e178c7b-riscv64-linux-gnu.tar.gz
f86ef652102f022827b70477bffa0a44008c6300cf62ca7b3595146cf2ed91ba  guix-build-1a0d8e178c7b/output/x86_64-apple-darwin/SHA256SUMS.part
f84d435d8e4709bf29bc7ac7ed8dc6b8af4077cef05e520b468b2896ce10876a  guix-build-1a0d8e178c7b/output/x86_64-apple-darwin/bitcoin-1a0d8e178c7b-x86_64-apple-darwin-unsigned.dmg
af2aab969b7ed7aeea0e02adbcc9e3b438086bf76b6bfc36146c53e05a27bd57  guix-build-1a0d8e178c7b/output/x86_64-apple-darwin/bitcoin-1a0d8e178c7b-x86_64-apple-darwin-unsigned.tar.gz
32a5109ba28ab74ff66238e6a8f8a04e455ebce382a3be287df92a227818fe72  guix-build-1a0d8e178c7b/output/x86_64-apple-darwin/bitcoin-1a0d8e178c7b-x86_64-apple-darwin.tar.gz
377462e9a96f4aba72c915dd5df5159a4301a1fa8ed0ee48faa6c71573de80c3  guix-build-1a0d8e178c7b/output/x86_64-linux-gnu/SHA256SUMS.part
a3bf62e828d2350a483b2d16205014f66e8884597b0b72e178042a958c548336  guix-build-1a0d8e178c7b/output/x86_64-linux-gnu/bitcoin-1a0d8e178c7b-x86_64-linux-gnu-debug.tar.gz
66cda980188ea1941a7d66c8b03c447580af33db55abe3bbe3581823ae0534a3  guix-build-1a0d8e178c7b/output/x86_64-linux-gnu/bitcoin-1a0d8e178c7b-x86_64-linux-gnu.tar.gz
2117f0dd9baeb4d585f841592e94c088f4487bf2008b8f281d0c3ceee92ff6cc  guix-build-1a0d8e178c7b/output/x86_64-w64-mingw32/SHA256SUMS.part
d40d5dec3287f467c42232c05d82f7fb538cda34bd2e63ff7e1876f471c3a790  guix-build-1a0d8e178c7b/output/x86_64-w64-mingw32/bitcoin-1a0d8e178c7b-win64-debug.zip
92dcc92765fbc07b1cc8258bfa69280541e1b4553cc41fed18672c2c6931d5c0  guix-build-1a0d8e178c7b/output/x86_64-w64-mingw32/bitcoin-1a0d8e178c7b-win64-setup-unsigned.exe
a6dd9b4d29f21d3a18cf64556cb03446ef17bf801eb6ac257b65d27cbd95080f  guix-build-1a0d8e178c7b/output/x86_64-w64-mingw32/bitcoin-1a0d8e178c7b-win64-unsigned.tar.gz
a4022e595d955198f73530473ef8e90a708746089ee2dd27de794176873330c1  guix-build-1a0d8e178c7b/output/x86_64-w64-mingw32/bitcoin-1a0d8e178c7b-win64.zip
```

luke-jr committed:
configure: Detect compatibility of Boost.Process rather than hardcode non-Windows

hebasto commented on configure.ac:
```diff
@@ -1476,45 +1476,33 @@ if test "$use_boost" = "yes"; then
+  TEMP_LDFLAGS="$LDFLAGS"
+  dnl Boost 1.73 and older require the following workaround.
+  LDFLAGS="$LDFLAGS $PTHREAD_CFLAGS"
+  AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <boost/process.hpp>]])],
```
Thank you @luke-jr! Cherry-picked your commit from #25111.

luke-jr replied:
Note there were 2 fixup commits as well. I think a7fa0da8611eb35ecd24284392f062a8ec47d037 might not be applicable here, but we probably need 06548df560f61c9deb5be09f72dd079833ec3c6b?

hebasto replied:
> Note there were 2 fixup commits as well. I think [a7fa0da](https://github.com/bitcoin/bitcoin/commit/a7fa0da8611eb35ecd24284392f062a8ec47d037) might not be applicable here

For sure, since v1.78.0. See https://github.com/boostorg/process/pull/171.

> but we probably need [06548df](https://github.com/bitcoin/bitcoin/commit/06548df560f61c9deb5be09f72dd079833ec3c6b)?

That issue has also been fixed since v1.78.0. See https://github.com/boostorg/process/pull/215.

hebasto replied:
Changes in `wchar_t.hpp` looks Windows-specific. FWIW, the first commit compiles with depends (boost 1.77) successfully.

@luke-jr Could you point at a configuration which fails?

fanquake commented:
[Similar to #25111. I'm still ~0 on this change](https://github.com/bitcoin/bitcoin/pull/25111#pullrequestreview-970602390), but this approach is an improvement. 

Given #24907, I don't think we should be leaning back into more Boost code usage. Outside of that, it's pretty clear that Boost Process currently has no or minimal support for mingw-w64 builds. It looks like they don't even have a mingw-w64 build [in their CI](https://drone.cpp.al/boostorg/process/87) (only MSVC). Windows fixes, some that we've upstreamed, also remain ignored, i.e https://github.com/boostorg/process/pull/237, https://github.com/boostorg/process/pull/239, https://github.com/boostorg/process/pull/240.

I assume if we make this change, external signing on Windows is only going to be supported with Boost 1.80.0 and later? If so, I assume we can remove https://github.com/bitcoin/bitcoin/blob/207a22877330709e4462e6092c265ab55c8653ac/src/test/system_tests.cpp#L10?

hebasto commented:
@fanquake 
> I assume if we make this change, external signing on Windows is only going to be supported with Boost 1.80.0 and later? If so, I assume we can remove
> 
> https://github.com/bitcoin/bitcoin/blob/207a22877330709e4462e6092c265ab55c8653ac/src/test/system_tests.cpp#L10

Correct. See #25723.

hebasto commented:
Rebased on top of #25723.

fanquake commented on depends/patches/boost/mingw32_process.patch:
```diff
@@ -0,0 +1,15 @@
+Use Boost.Process with mingw32
```
Is this patch from upstream? If so, can you add the upstream commit details / links here? If not, are you planning on upstreaming it? The fact that we still have to patch Process out the box, just to make it work, isn't encouraging. Maybe these fixes can make it into 1.80.0.

hebasto replied:
> If not, are you planning on upstreaming it?

https://github.com/boostorg/process/pull/264

Sjors commented:
I can't test on Windows. Bumping the Boost version in depends should probably be its own PR? (or at least indicated in the PR description)

fanquake commented:
> Bumping the Boost version in depends should probably be its own PR?

See #25873.

hebasto commented:
Updated 0444353ba7e97b53ff7f4ed7f06d67ed36466912 -> 9077916135ddabc07f105a77da8eb2bf8de73469 ([pr25696.04](https://github.com/hebasto/bitcoin/commits/pr25696.04) -> [pr25696.05](https://github.com/hebasto/bitcoin/commits/pr25696.05)):

- rebased on top of the #25873
- a patch replaced with the [upstream one](https://github.com/boostorg/process/pull/264)
- undrafted

jarolrod commented:
concept ack

hebasto commented:
Rebased 9077916135ddabc07f105a77da8eb2bf8de73469 -> 75b51ed60350d63e1216be21f345bb297e697374 ([pr25696.05](https://github.com/hebasto/bitcoin/commits/pr25696.05) -> [pr25696.06](https://github.com/hebasto/bitcoin/commits/pr25696.06)) on top of the merged #25873.

Sjors commented:
I may have a Windows dual boot system in a month or so, so I can test stuff like this.

jarolrod reviewed (approved):
ACK 75b51ed60350d63e1216be21f345bb297e697374

I tested that external signer works again on windows. The external signer setting is now show in the GUI. I was able to connect an external signer, communicate with HWI, and create a new wallet. I confirmed the correct descriptors were passed, I was able to generate new addresses, receive testnet bitcoin, and sign & broadcast a transaction.

Sjors commented on depends/patches/boost/mingw32_process.patch:
```diff
@@ -0,0 +1,21 @@
+Fix compiling for MinGW-w64 using std::filesystem 
+
+Upstream commit:
+ - https://github.com/boostorg/process/commit/a7b65bfc44f4af3a9d584fe9347fde0944c0167a
```
This has been merged. Do we know which release it's in?

hebasto replied:
[Updated](https://github.com/bitcoin/bitcoin/pull/25696#issuecomment-1324936336).

Sjors reviewed (commented):
Testing 75b51ed60350d63e1216be21f345bb297e697374 rebased on master...

External signer works on Windows 11 when using the Guix build. I can also see that the it won't build the external signer if I drop the last commit.

The patch matches what's upstreamed to Boost.

However, in PowerShell as well the regular Command Prompt the tests fail:

```
 .\test_bitcoin.exe
Running 534 test cases...
test/system_tests.cpp(66): error: in "system_tests/run_command": check what.find(expected) != std::string::npos has failed
test/system_tests.cpp(99): error: in "system_tests/run_command": check what.find(expected) != std::string::npos has failed
```

hebasto commented:
@Sjors 
> However, in PowerShell as well the regular Command Prompt the tests fail:

Is your Windows localized? See #26368.

Sjors commented:
It is, I'll try that fix.

hebasto commented:
Rebased 75b51ed60350d63e1216be21f345bb297e697374 -> 8db9f3c4bd1e4ebad3b89738544b38f103a7690a ([pr25696.06](https://github.com/hebasto/bitcoin/commits/pr25696.06) -> [pr25696.07](https://github.com/hebasto/bitcoin/commits/pr25696.07)) on top of the merged #26377.

hebasto commented:
Rebased 75b51ed60350d63e1216be21f345bb297e697374 -> 86cb0b7168a2e011e5e2f6ac2a605db64b0525ba ([pr25696.06](https://github.com/hebasto/bitcoin/commits/pr25696.06) -> [pr25696.08](https://github.com/hebasto/bitcoin/commits/pr25696.08)) on top of the merged #26377.

Sjors commented:
Tested 86cb0b7168a2e011e5e2f6ac2a605db64b0525ba using guix build. External signer works, but tests fail:

```
.\bitcoin-86cb0b7168a2\bin\test_bitcoin.exe
Running 534 test cases...
test/system_tests.cpp(61): error: in "system_tests/run_command": check e.code().value() == expected_error has failed [2 != 6]
test/system_tests.cpp(94): error: in "system_tests/run_command": check what.find(expected) != std::string::npos has failed

*** 2 failures are detected in the test module "Bitcoin Core Test Suite"
```

hebasto commented:
Updated 86cb0b7168a2e011e5e2f6ac2a605db64b0525ba -> 3e8160b1a27bda6ca41f2b8a577e5a7764ebefae ([pr25696.08](https://github.com/hebasto/bitcoin/commits/pr25696.08) -> [pr25696.09](https://github.com/hebasto/bitcoin/commits/pr25696.09)):

- rebased on top of the #26557 to avoid Boost source patching
- fixed tests

@Sjors https://github.com/bitcoin/bitcoin/pull/25696#issuecomment-1297219663 should be fixed now.

hebasto commented on src/test/system_tests.cpp:
```diff
@@ -66,20 +70,25 @@ BOOST_AUTO_TEST_CASE(run_command)
         // Return non-zero exit code, with error message for stderr
 #ifdef WIN32
         const std::string command{"cmd.exe /c dir nosuchfile"};
+#ifdef _MSC_VER
```
No [longer](https://github.com/bitcoin/bitcoin/pull/25696#issuecomment-1324936336) relevant, I guess.

hebasto committed:
build: Re-enable external signer on Windows

hebasto commented:
Rebased bfd8f07293c66afdd5e8b91f99724e2d95a2ba36 -> 1a0d8e178c7b9a3e94d14f94b77305802ebdc93b ([pr25696.10](https://github.com/hebasto/bitcoin/commits/pr25696.10) -> [pr25696.11](https://github.com/hebasto/bitcoin/commits/pr25696.11)) on top of the #26557.

Ready for (final?) reviewing :)

Sjors commented:
Tests pass now.

tACK 1a0d8e178c7b9a3e94d14f94b77305802ebdc93b

Also checked that with 989451d0689543b25ee6bf1d5b82c863d583597b Boost:Process is still detected correctly on Intel macOS.

hebasto commented:
@achow101 @luke-jr @jarolrod 

Mind looking into this PR?

achow101 commented:
ACK 1a0d8e178c7b9a3e94d14f94b77305802ebdc93b

fanquake commented:
post-merge Concept NACK. I don't know why we are leaning back into this.

hebasto commented:
> I don't know why we are leaning back into this.

Because using external signers is one of the best security options for users who run their Bitcoin Core software on Windows.

> post-merge Concept NACK. 

Better alternatives are welcome.
